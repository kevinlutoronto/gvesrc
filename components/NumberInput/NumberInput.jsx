// @flow
import { toNumber, toString } from 'lodash';
import React from 'react';
import cx from 'classnames';
import style from './style.scss';
import { getDataAttrs } from '../../dataUtils';
import type { Data } from '../../types';

type HandleEvent = {|
  target: {
    parentElement: {
      classList: {
        remove: string => mixed,
        add: string => mixed,
      },
    },
  },
  nativeEvent: {
    target: {
      value: Object,
    },
  },
|};

type Props = {|
  data?: Data,
  disabled?: boolean,
  error?: boolean,
  maxValue: number,
  minValue: number,
  onBlur?: HandleEvent => void,
  onChange: HandleEvent => void,
  onFocus?: HandleEvent => void,
  placeholder: string,
  step: number,
  value: string,
|};

class NumberInput extends React.Component<Props> {
  static defaultProps = {
    value: '',
    placeholder: '',
    step: 1,
    maxValue: Number.MAX_SAFE_INTEGER,
    minValue: Number.MIN_SAFE_INTEGER,
    onChange() {},
  };

  handleIncrement = (): void => {
    const { maxValue, value, step, onChange } = this.props;
    const prevValue = toNumber(value);
    const nextValue = prevValue + step;
    if (maxValue >= nextValue) onChange(toString(nextValue));
  };

  handleDecrement = (): void => {
    const { minValue, value, step, onChange } = this.props;
    const prevValue = toNumber(value);
    const nextValue = prevValue - step;
    if (minValue <= nextValue) onChange(toString(nextValue));
  };

  handleChange = (e: HandleEvent): void => {
    this.props.onChange && this.props.onChange(e.nativeEvent.target.value);
  };

  handleBlur = (e: HandleEvent): void => {
    e.target.parentElement.classList.remove(style.wrapperFocused);
    this.props.onBlur && this.props.onBlur(e.nativeEvent.target.value);
  };

  handleFocus = (e: HandleEvent): void => {
    e.target.parentElement.classList.add(style.wrapperFocused);
    this.props.onFocus && this.props.onFocus(e.nativeEvent.target.value);
  };

  render() {
    const { disabled, error, data, placeholder, value, step, minValue, maxValue } = this.props;

    const wrapperStyle = cx(style.wrapper, {
      [style.wrapperDisabled]: disabled,
      [style.wrapperError]: error,
    });

    const spinnerStyle = cx(style.spinner, {
      [style.spinnerDisabled]: disabled,
    });

    const inputStyle = cx(style.input, {
      [style.inputDisabled]: disabled,
    });

    const actionStyle = cx(style.action, {
      [style.buttonDisabled]: disabled,
    });

    return (
      <div {...getDataAttrs(data)} className={wrapperStyle}>
        <input
          type="number"
          placeholder={placeholder}
          className={inputStyle}
          disabled={disabled}
          value={value}
          step={step}
          min={minValue}
          max={maxValue}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
          onFocus={this.handleFocus}
        />
        <div className={spinnerStyle}>
          <div className={actionStyle} onClick={!disabled && this.handleIncrement}>
            +
          </div>
          <div className={actionStyle} onClick={!disabled && this.handleDecrement}>
            -
          </div>
        </div>
      </div>
    );
  }
}

NumberInput.displayName = 'Plasma@NumberInput';

export default NumberInput;
