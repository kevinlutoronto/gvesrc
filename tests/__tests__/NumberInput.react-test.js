/* eslint-disable */

import React from 'react';
import NumberInput from '../../components/NumberInput/NumberInput';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

describe('NumberInput Component', () => {
  test('render', () => {
    const component = renderer.create(<NumberInput onChange={() => {}} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
