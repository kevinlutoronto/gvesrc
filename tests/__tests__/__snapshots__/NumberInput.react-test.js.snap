// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput Component render 1`] = `
<div
  className="wrapper"
>
  <input
    className="input"
    disabled={undefined}
    max={9007199254740991}
    min={-9007199254740991}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    step={1}
    type="number"
    value=""
  />
  <div
    className="spinner"
  >
    <div
      className="action"
      onClick={[Function]}
    >
      +
    </div>
    <div
      className="action"
      onClick={[Function]}
    >
      -
    </div>
  </div>
</div>
`;
